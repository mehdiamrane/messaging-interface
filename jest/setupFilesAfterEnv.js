import '@testing-library/jest-dom/extend-expect';

jest.mock('next-i18next', () => ({
  useTranslation: () => ({ t: (key) => key }),
}));
