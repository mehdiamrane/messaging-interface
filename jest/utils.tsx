import React from 'react';
import { render, type RenderOptions, type RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import theme from 'src/styles/theme';

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const AllProviders = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

interface Render {
  (ui: React.ReactElement, options?: RenderOptions): RenderResult;
}

const customRender: Render = (ui, options) =>
  render(ui, {
    wrapper: (props) => <AllProviders {...props} />,
    ...options,
  });

export * from '@testing-library/react';
export { customRender as render };
