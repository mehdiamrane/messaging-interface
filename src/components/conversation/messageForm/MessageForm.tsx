import React, { FC } from 'react';
import { StyledMessageForm } from './MessageForm.style';
import { useTranslation } from 'next-i18next';
import { HiPaperAirplane } from 'react-icons/hi';

type MessageFormProps = {
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  value: string;
  onSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
};

const MessageForm: FC<MessageFormProps> = ({ onChange, value, onSubmit }) => {
  const { t } = useTranslation('common');

  return (
    <StyledMessageForm.Container onSubmit={onSubmit}>
      <StyledMessageForm.Input
        required
        value={value}
        onChange={onChange}
        minLength={1}
        placeholder={t('conversation.form.input.placeholder')}
      />
      <StyledMessageForm.Button type='submit' title={t('conversation.form.button.title')}>
        <StyledMessageForm.ButtonText>{t('conversation.form.button.label')}</StyledMessageForm.ButtonText>
        <HiPaperAirplane size='1.25rem' />
      </StyledMessageForm.Button>
    </StyledMessageForm.Container>
  );
};

export default MessageForm;
