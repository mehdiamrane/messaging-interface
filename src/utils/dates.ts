import formatDistanceToNow from 'date-fns/formatDistanceToNow';
import intlFormat from 'date-fns/intlFormat';
import { enUS, fr } from 'date-fns/locale';

const localeMapping = { 'en-US': enUS, fr };

export const formatRelativeDate = (date: Date, locale: string): string =>
  formatDistanceToNow(date, { locale: localeMapping[locale], addSuffix: true });

export const formatDate = (date: Date, locale: string): string =>
  intlFormat(
    date,
    {
      weekday: 'long',
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
    },
    {
      locale,
    },
  );

export const timestampToDate = (timestamp: number): Date => new Date(timestamp * 1000);
