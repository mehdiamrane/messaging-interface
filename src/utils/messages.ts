import type { Message } from 'src/types/message';
import type { ChatBubbleProps } from 'src/components/conversation/chatBubble/ChatBubble';

export const sortMessages = (messages: Message[]): Message[] =>
  (messages || []).sort((a, b) => a.timestamp - b.timestamp);

export const getMessagePosition = (messages: Message[], index: number): ChatBubbleProps['position'] => {
  const previous = messages[index - 1];
  const current = messages[index];
  const next = messages[index + 1];
  if (previous?.authorId !== current.authorId && next?.authorId !== current.authorId) {
    return 'single';
  }
  if (previous?.authorId !== current.authorId && next?.authorId === current.authorId) {
    return 'first';
  }
  if (previous?.authorId === current.authorId && next?.authorId !== current.authorId) {
    return 'last';
  }
  return 'middle';
};
