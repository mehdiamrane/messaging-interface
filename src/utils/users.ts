import type { User } from 'src/types/user';

// Default way to use a logged user
export const getLoggedUserId = (): User['id'] => 1;

export const isLoggedUser = (userId: User['id']): boolean => userId === getLoggedUserId();

export const getUserData = (userId: User['id'], users: User[]): User | undefined =>
  users.find(({ id }) => id === userId);

export const getAvatarFromNickname = (nickname: User['nickname']): string =>
  `https://ui-avatars.com/api/?bold=true&size=54&background=CBD5E0&rounded=true&name=${nickname}`;
