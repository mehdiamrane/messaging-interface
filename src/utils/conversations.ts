import type { Conversation } from 'src/types/conversation';
import type { BareUser } from 'src/types/user';
import { getLoggedUserId } from './users';

export const sortConversationsByLastActive = (conversations: Conversation[]): Conversation[] =>
  conversations.sort((a, b) => b.lastMessageTimestamp - a.lastMessageTimestamp);

export const getOtherUserFromConversation = (conversation: Conversation): BareUser =>
  conversation.senderId === getLoggedUserId()
    ? { id: conversation.recipientId, nickname: conversation.recipientNickname }
    : { id: conversation.senderId, nickname: conversation.senderNickname };
