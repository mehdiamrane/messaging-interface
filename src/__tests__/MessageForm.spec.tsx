import React from 'react';
import { render } from 'jest/utils';
import MessageForm from 'src/components/conversation/messageForm/MessageForm';
import 'jest-styled-components';

describe('MessageForm', () => {
  it('should render correctly MessageForm', () => {
    const { container } = render(<MessageForm onChange={jest.fn()} onSubmit={jest.fn()} value='' />);
    expect(container).toMatchSnapshot();
  });
});
