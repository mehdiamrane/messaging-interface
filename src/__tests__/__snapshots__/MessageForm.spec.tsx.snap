// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm should render correctly MessageForm 1`] = `
.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.75rem;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5rem;
  width: 100%;
}

.c1 {
  border: 2px solid #CBD5E0;
  border-radius: 0.375rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  line-height: 1.5;
  font-size: 1rem;
  padding: 0.5rem;
  resize: none;
  -webkit-transition: border 150ms ease;
  transition: border 150ms ease;
}

.c1:hover {
  border: 2px solid #A0AEC0;
}

.c1:focus {
  border: 2px solid #63b3ed;
  outline: none;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #90cdf4;
  border: 2px solid #63b3ed;
  border-radius: 0.375rem;
  color: #1A365D;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1rem;
  font-weight: 600;
  padding: 0 0.75rem;
  -webkit-transition: opacity 150ms ease,border 150ms ease,background-color 150ms ease;
  transition: opacity 150ms ease,border 150ms ease,background-color 150ms ease;
}

.c2:hover {
  background-color: #63b3ed;
  border: 2px solid #4299e1;
}

.c2:active {
  background-color: #bee3f8;
  border: 2px solid #90cdf4;
}

.c2[aria-disabled='true'] {
  opacity: 0.5;
  cursor: not-allowed;
}

.c2[aria-disabled='true']:active {
  pointer-events: none;
}

.c3 {
  display: none;
}

@media (min-width:48em) {
  .c3 {
    display: initial;
    margin-right: 0.5rem;
  }
}

<div>
  <form
    class="c0"
  >
    <input
      class="c1"
      minlength="1"
      placeholder="conversation.form.input.placeholder"
      required=""
      value=""
    />
    <button
      class="c2"
      title="conversation.form.button.title"
      type="submit"
    >
      <span
        class="c3"
      >
        conversation.form.button.label
      </span>
      <svg
        fill="currentColor"
        height="1.25rem"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1.25rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.894 2.553a1 1 0 00-1.788 0l-7 14a1 1 0 001.169 1.409l5-1.429A1 1 0 009 15.571V11a1 1 0 112 0v4.571a1 1 0 00.725.962l5 1.428a1 1 0 001.17-1.408l-7-14z"
        />
      </svg>
    </button>
  </form>
</div>
`;
